cpp/src/cider/exec/nextgen/operators/RowToColumnNode.cpp

```diff
@@ -127,6 +132,178 @@ class ColumnWriter {
         setNullBuffer(values.getNull(), for_null), raw_length_buffer, raw_data_buffer);
   }
 
+  // allocate and copy 4 buffers:
+  // 1.validity bitmap buffer(represent whether the row is null) in parent arrow array
+  // 2.offsets buffer in parent
+  // 3.validity bitmap buffer(represent whether the element is null) in child arrow array
+  // 4.actual values buffer in child
+  void writeVariableSizeArrayCol(bool for_null) {
+    // Get values need to write
+    utils::VarSizeArrayExprValue values(expr_->get_expr_value());
+
+    if (for_null) {
+      auto null = setNullBuffer(values.getNull(), for_null);
+      Analyzer::JITExprValueAdaptor(
+          context_.getArrowArrayValues(expr_->getLocalIndex()).second)
+          .setNull(null);
+      return;
+    }
+
+    // allocate offset buffer, need array_len + 1 element.
+    auto new_offset_buffer = context_.getJITFunction()->createLocalJITValue([this]() {
+      auto offsets_bytes_num =
+          (arrow_array_len_ + 1) *
+          context_.getJITFunction()->createLiteral(JITTypeTag::INT64, 4);
+      return allocateRawDataBuffer(1, offsets_bytes_num);
+    });
+    auto new_offset_bufferi32 = new_offset_buffer->castPointerSubType(JITTypeTag::INT32);
+
+    // allocate validity bitmap in child array
+    // reallocate buffer on demand
+    auto child_array = context_.getJITFunction()->createLocalJITValue(
+        [this]() { return context::codegen_utils::getArrowArrayChild(arrow_array_, 0); });
+    auto bitmap_bytes_len = (values.getOffset() + values.getLength() + 7) / 8;
+    auto new_elem_null_buffer = context_.getJITFunction()->emitRuntimeFunctionCall(
+        "get_data_buffer_with_realloc_on_demand",
+        JITFunctionEmitDescriptor{
+            .ret_type = JITTypeTag::POINTER,
+            .ret_sub_type = JITTypeTag::INT8,
+            .params_vector = {
+                child_array.get(),
+                bitmap_bytes_len->castJITValuePrimitiveType(JITTypeTag::INT32).get(),
+                context_.getJITFunction()->createLiteral(JITTypeTag::INT32, 0).get()}});
+
+    // allocate values buffer in child
+    auto values_bytes_len =
+        (values.getLength() + values.getOffset()) *
+        utils::getTypeBytes(expr_->get_type_info().getChildAt(0).get_type());
+    auto raw_data_buffer = context_.getJITFunction()->emitRuntimeFunctionCall(
+        "get_data_buffer_with_realloc_on_demand",
+        JITFunctionEmitDescriptor{
+            .ret_type = JITTypeTag::POINTER,
+            .ret_sub_type = JITTypeTag::INT8,
+            .params_vector = {
+                child_array.get(),
+                values_bytes_len->castJITValuePrimitiveType(JITTypeTag::INT32).get(),
```

Comment:
This parameter is given as "current bytes ALREADY wrote in buffer", but here you give "current bytes ALREADY wrote in buffer + current row data length"?